dialog/dialog.1
.TH DIALOG 1 "10 June 1994" "Dialog Version 0.6"
.SH NAME
dialog \- display dialog boxes from shell scripts
.SH SYNOPSIS
.B dialog --clear
.br
.BI "dialog --create-rc " file
.br
.B dialog
[
.B \-\-title
.I title
]
[
.B \-\-backtitle
.I backtitle
]
[
.B \-\-clear
]
[
.B \-\-separate\-output
]
.B box-options
.SH DESCRIPTION
.B Dialog
is a program that will let you to present a variety of questions or
display messages using dialog boxes from a shell script. Currently,
these types of dialog boxes are implemented: 
.LP
.BR yes/no " box," " menu" " box," " input" " box,"
.BR message " box," " text" " box," " info" " box,"
.BR checklist " box," " radiolist" " box, and" " gauge" " box."
.SH OPTIONS
.TP
.B \-\-clear
The screen will be cleared to the
.BR "screen attribute" " on exit."
.TP
.BI \-\-create-rc " file"
.RB "Since " dialog " supports run-time configuration,"
this can be used to dump a sample configuration file to the file specified
by
.IR file "."
.TP
.BI \-\-separate\-output
For checklist widgets, output result one line at a time, with no
quoting.  This facilitates parsing by another program.
.TP
.BI \-\-title " title"
Specifies a
.I title
string to be displayed at the top of the dialog box.
.TP
.BI \-\-backtitle " backtitle"
Specifies a
.I backtitle
string to be displayed on the backdrop, at the top of the screen.
.TP
.B Box Options
.TP
.BI \-\-yesno " text height width"
.RB A " yes/no" " dialog box of size"
.I height
rows by
.I width
columns will be displayed. The string specified by
.I text
is displayed inside the dialog box. If this string is too long to be fitted
in one line, it will be automatically divided into multiple lines at
appropriate places. The
.I text
string may also contain the sub-string
.I
"\en"
or newline characters
.I `\en\'
to control line breaking explicitly.  This dialog box is useful for
asking questions that require the user to answer either yes or no.
.RB "The dialog box has a" " Yes" " button and a " No
button, in which the user can switch between by pressing the
.IR TAB " key."
.TP
.BI \-\-msgbox " text height width"
.RB A " message" " box is very similar to a" " yes/no" " box."
The only difference between a
.B message
box and a
.B yes/no
box is that a
.B message
box has only a single
.B OK
button. You can use this dialog box to display any message you like.
After reading the message, the user can press the
.I ENTER
key so that
.B dialog
will exit and the calling shell script can continue its operation.
.TP
.BI \-\-infobox " text height width"
.RB An " info" " box is basically a" " message" " box."
However, in this case,
.B dialog
will exit immediately after displaying the message to the user. The
screen is not cleared when
.B dialog
exits, so that the message will remain on the screen until the calling
shell script clears it later. This is useful when you want to inform
the user that some operations are carrying on that may require some
time to finish.
.TP
.BI \-\-inputbox " text height width [init]"
.RB "An " input " box is useful when you want to ask questions that"
require the user to input a string as the answer. If init is supplied
it is used to initialize the input string.
When inputing the
string, the
.I BACKSPACE
key can be used to correct typing errors. If the input string is longer than
can be fitted in the dialog box, the input field will be scrolled. On exit,
the input string will be printed on
.IR stderr "."
.TP
.BI \-\-textbox " file height width"
.RB A " text" " box lets you display the contents of a text file in a"
dialog box. It is like a simple text file viewer. The user can move
through the file by using the
.IR UP/DOWN ", " PGUP/PGDN
.RI and " HOME/END" " keys available on most keyboards."
If the lines are too long to be displayed in the box, the
.I LEFT/RIGHT
keys can be used to scroll the text region horizontally. For more
convenience, forward and backward searching functions are also provided.
.IP "\fB\-\-menu \fItext height width menu-height \fR[ \fItag item \fR] \fI..."
As its name suggests, a
.B menu
box is a dialog box that can be used to present a list of choices in
the form of a menu for the user to choose. Each menu entry consists of a
.IR tag " string and an " item " string. The"
.I tag
gives the entry a name to distinguish it from the other entries in the
menu. The
.I item
is a short description of the option that the entry represents. The
user can move between the menu entries by pressing the
.I UP/DOWN
keys, the first letter of the
.I tag
as a hot-key, or the number keys
.IR 1-9 ". There are"
.I menu-height
entries displayed in the menu at one time, but the menu will be
scrolled if there are more entries than that. When
.B dialog
exits, the
.I tag
of the chosen menu entry will be printed on
.IR stderr "."
.IP "\fB\-\-checklist \fItext height width list-height \fR[ \fItag item status \fR] \fI..."
.RB "A " checklist " box is similar to a " menu " box in that there are"
multiple entries presented in the form of a menu. Instead of choosing
one entry among the entries, each entry can be turned on or off by the
user. The initial on/off state of each entry is specified by
.IR status "."
On exit, a list of the
.I tag
strings of those entries that are turned on will be printed on
.IR stderr "."

.IP "\fB\-\-radiolist \fItext height width list-height \fR [ \fItag item status \fR] \fI..."
.RB "A " radiolist " box is similar to a " menu " box.  The only difference is"
that you can indicate which entry is currently selected, by setting its
.IR status " to " on "."

.IP "\fB\-\-gauge \fItext height width percent\fR"
.RB "A " gauge " box displays a meter along the bottom of the box.
The meter indicates the percentage.  New percentages are read from
standard input, one integer per line.  The meter is updated
to reflect each new percentage.  If stdin is XXX, then subsequent
lines up to another XXX are used for a new prompt.
The gauge exits when EOF is reached on stdin.

.SH "RUN-TIME CONFIGURATION"
.TP 4
1.
Create a sample configuration file by typing:
.LP
.in +1i
"dialog --create-rc <file>"
.TP 4
2.
At start,
.B dialog
determines the settings to use as follows:
.RS
.TP 4
a)
if environment variable
.B DIALOGRC
is set, it's value determines the name of the configuration file.
.TP 4
b)
if the file in (a) can't be found, use the file
.I $HOME/.dialogrc
as the configuration file.
.TP 4
c)
if the file in (b) can't be found, use compiled in defaults.
.RE
.TP 4
3.
Edit the sample configuration file and copy it to some place that
.B dialog
can find, as stated in step 2 above.
.SH ENVIROMENT
.TP 15
.B DIALOGRC
Define this variable if you want to specify the name of the configuration file
to use.
.SH FILES
.TP 20
.I $HOME/.dialogrc
default configuration file
.SH DIAGNOSTICS
Exit status is 0 if
.BR dialog " is exited by pressing the " Yes " or " OK
button, and 1 if the
.BR No " or " Cancel
button is pressed. Otherwise, if errors occur inside
.B dialog
or
.B dialog
is exited by pressing the
.I ESC
key, the exit status is -1.
.SH BUGS
Text files containing
.I tab
characters may cause problems with
.B text
box.
.I Tab
characters in text files must first be expanded to spaces before being
.RB "displayed by " text " box."
.TP
Screen update is too slow.
.SH AUTHOR
Savio Lam (lam836@cs.cuhk.hk) - version 0.3
.LP
Stuart Herbert (S.Herbert@sheffield.ac.uk) - patch for version 0.4


CHANGES

3.1.6-109.75-1
--------------
- adapted to new kernels 3.x
- update driver Divatty.ko
- scripts need bash

3.1.5-109.75-1
--------------
- adapted to new kernel 2.6.34

3.1.4-109.75-1
--------------
- updated firmware
- minor line-interconnect changes

3.1.3-109.50-1
--------------
- adapted to new PROC API changes of kernel 2.6.30

3.1.2-109.50-1
--------------
- re-activated older cards

3.1.1-109.50-1
--------------
- Fixed cross board line interconnect
- Support for new PCIe cards (PRI, 2PRI, 4PRI)
- Added TTY driver (ttyds nodes with AT commands)

3.1.0-109.42-1
--------------
- Complete update to new Dialogic 9.0 package
- removed support for kernel 2.4.x
- removed support for older cards

3.0.11-107.884-1
----------------
- Adaptions for kernels until 2.6.28

3.0.10-107.884-1
----------------
- Adaptions for kernels 2.6.22
- Support for new PCIe cards.
- Fixed /proc path.
- Minor feature fixes.
- Firmware update.

3.0.9-107.775-1
---------------
- fixed install of non-existent kernelcapi modules with kernel 2.6
- firmware update
- kernel 2.4 driver update
- copyright/vendor changes

3.0.8-107.755-1
---------------
- firmware update
- divactrl capi-management debug
- kernel 2.6.20 adaptions
- copyright/vendor changes 

3.0.7-107.725-1
---------------
- Removed warning on card startup about missing index.html

3.0.6-107.725-1
---------------
- Adaptions for kernel 2.6.19
- Newest available CAPI common code
- New firmware to fix 4BRI Rev. 1 startup
- One or more ports of 4BRI Rev. 2 adapter can be used for 1TR6 in the time other ports are used with other protocols.
  * Configuration automatically uses V.14 protocol code in this case.
    Please note that use of V.14 protocol code introduces following limitation:
    + Cross board line interconnect is not compatible with V.15 protocol code or newer
    + Only set of features supported by V.14 protocol code is provided



dialog/dialog.c
/*
 *  dialog - Display simple dialog boxes from shell scripts
 *
 *  AUTHOR: Savio Lam (lam836@cs.cuhk.hk)
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "dialog.h"

static void Usage (const char *name);

static int separate_output = 0;

typedef int (jumperFn) (const char *title, int argc, const char * const * argv);

struct Mode {
    char *name;
    int argmin, argmax, argmod;
    jumperFn *jumper;
};

jumperFn j_yesno, j_msgbox, j_infobox, j_textbox, j_menu;
jumperFn j_checklist, j_radiolist, j_inputbox, j_gauge;

/*
 * All functions are used in the slackware root disk, apart from "gauge"
 */

static struct Mode modes[] =
{
    {"--yesno", 5, 5, 1, j_yesno},
    {"--msgbox", 5, 5, 1, j_msgbox},
    {"--infobox", 5, 5, 1, j_infobox},
    {"--textbox", 5, 5, 1, j_textbox},
    {"--menu", 8, 0, 2, j_menu},
    {"--checklist", 9, 0, 3, j_checklist},
    {"--radiolist", 9, 0, 3, j_radiolist},
    {"--inputbox", 5, 6, 1, j_inputbox},
#ifdef HAVE_GAUGE
    {"--gauge", 6, 6, 1, j_gauge},
#endif
    {NULL, 0, 0, 0, NULL}
};

static struct Mode *modePtr;

#ifdef LOCALE
#include <locale.h>
#endif

int
main (int argc, const char * const * argv)
{
    int offset = 0, clear_screen = 0, end_common_opts = 0, retval;
    const char *title = NULL;

#ifdef LOCALE
    (void) setlocale (LC_ALL, "");
#endif

    if (argc < 2) {
	Usage (argv[0]);
	exit (-1);
    }
#ifdef HAVE_RC_FILE

    else if (!strcmp (argv[1], "--create-rc")) {
#ifdef HAVE_NCURSES
	if (argc != 3) {
	    Usage (argv[0]);
	    exit (-1);
	}
	_dialog_create_rc (argv[2]);
	return 0;
#else
	fprintf (stderr, "This option is currently unsupported "
		 "on your system.\n");
	return -1;
#endif
    }
#endif

    while (offset < argc - 1 && !end_common_opts) {	/* Common options */
	if (!strcmp (argv[offset + 1], "--title")) {
	    if (argc - offset < 3 || title != NULL) {
		Usage (argv[0]);
		exit (-1);
	    } else {
		title = argv[offset + 2];
		offset += 2;
	    }
	} else if (!strcmp (argv[offset + 1], "--backtitle")) {
	    if (backtitle != NULL) {
		Usage (argv[0]);
		exit (-1);
	    } else {
		backtitle = argv[offset + 2];
		offset += 2;
	    }
	} else if (!strcmp (argv[offset + 1], "--separate-output")) {
	    separate_output = 1;
	    offset++;
	} else if (!strcmp (argv[offset + 1], "--clear")) {
	    if (clear_screen) {	/* Hey, "--clear" can't appear twice! */
		Usage (argv[0]);
		exit (-1);
	    } else if (argc == 2) {	/* we only want to clear the screen */
		dialog_init ();
		refresh ();	/* dialog_init() will clear the screen for us */
		dialog_end ();
		return 0;
	    } else {
		clear_screen = 1;
		offset++;
	    }
	} else if (!strcmp (argv[offset + 1], "--helptag")) {
		helptag = argv[offset + 2];
		offset += 2;
	} else			/* no more common options */
	    end_common_opts = 1;
    }

    if (argc - 1 == offset) {	/* no more options */
	Usage (argv[0]);
	exit (-1);
    }
    /* use a table to look for the requested mode, to avoid code duplication */

    for (modePtr = modes; modePtr->name; modePtr++)	/* look for the mode */
	if (!strcmp (argv[offset + 1], modePtr->name))
	    break;

    if (!modePtr->name)
	Usage (argv[0]);
    if (argc - offset < modePtr->argmin)
	Usage (argv[0]);
    if (modePtr->argmax && argc - offset > modePtr->argmax)
	Usage (argv[0]);
    if ((argc - offset) % modePtr->argmod)
	Usage (argv[0]);

    dialog_init ();
    retval = (*(modePtr->jumper)) (title, argc - offset, argv + offset);

    if (clear_screen) {		/* clear screen before exit */
	_dialog_attr_clear (stdscr, LINES, COLS, screen_attr);
	refresh ();
    }
    dialog_end ();

    exit (retval);
}

/*
 * Print program usage
 */
static void
Usage (const char *name)
{
    fprintf (stderr, "\
\ndialog version " DIALOG_VERSION ".\
\nDisplay dialog boxes from shell scripts *\
\n\
\nUsage: %s --clear\
\n       %s --create-rc <file>\
\n       %s [--title <title>] [--separate-output] [--backtitle <backtitle>] [--clear] <Box options>\
\n\
\nBox options:\
\n\
\n  --helptag   <file>\
\n  --yesno     <text> <height> <width>\
\n  --msgbox    <text> <height> <width>\
\n  --infobox   <text> <height> <width>\
\n  --inputbox  <text> <height> <width> [<init>]\
\n  --textbox   <file> <height> <width>\
\n  --menu      <text> <height> <width> <menu height> <tag1> <item1>...\
\n  --checklist <text> <height> <width> <list height> <tag1> <item1> <status1>...\
\n  --radiolist <text> <height> <width> <list height> <tag1> <item1> <status1>...\
\n  --gauge     <text> <height> <width> <percent>\n", name, name, name);
    exit (-1);
}

/*
 * These are the program jumpers
 */

int
j_yesno (const char *t, int ac, const char * const * av)
{
    return dialog_yesno (t, av[2], atoi (av[3]), atoi (av[4]));
}

int
j_msgbox (const char *t, int ac, const char * const * av)
{
    return dialog_msgbox (t, av[2], atoi (av[3]), atoi (av[4]), 1);
}

int
j_infobox (const char *t, int ac, const char * const * av)
{
    return dialog_msgbox (t, av[2], atoi (av[3]), atoi (av[4]), 0);
}

int
j_textbox (const char *t, int ac, const char * const * av)
{
    return dialog_textbox (t, av[2], atoi (av[3]), atoi (av[4]));
}

int
j_menu (const char *t, int ac, const char * const * av)
{
    int ret = dialog_menu (t, av[2], atoi (av[3]), atoi (av[4]),
			atoi (av[5]), (ac - 6) / 2, av + 6);
    if (ret >= 0) {
	fprintf(stderr, av[6 + ret * 2]);
	return 0;
    } else if (ret == -2)
	return 1;	/* CANCEL */
    return ret;		/* ESC pressed, ret == -1 */
}

int
j_checklist (const char *t, int ac, const char * const * av)
{
    int item_no = (ac - 6) / 3;
    int i, ret, *status = malloc (sizeof (int) * item_no);
    if (status == NULL) {
	endwin ();
	fprintf (stderr, "\nCan't allocate memory in dialog_checklist().\n");
	exit (-1);
    }
    do {
        ret = dialog_checklist (t, av[2], atoi (av[3]), atoi (av[4]),
	    atoi (av[5]), item_no, av + 6, FLAG_CHECK, status);
        if (helptag && ret == 2) {
            dialog_clear();
            dialog_textbox("Help", helptag, 18, 70);
            dialog_clear();
        }
    } while (ret == 2);
    if (ret == 0) {
	for (i = 0; i < item_no; i++)
	    if (status[i]) {
		if (separate_output)
		    fprintf (stderr, "%s\n", (av + 6)[i * 3]);
		else
		    fprintf (stderr, "\"%s\" ", (av + 6)[i * 3]);
	    }
    }
    free (status);
    return ret;
}

int
j_radiolist (const char *t, int ac, const char * const * av)
{
    int item_no = (ac - 6) / 3;
    int i, ret, *status = malloc (sizeof (int) * item_no);
    if (status == NULL) {
	endwin ();
	fprintf (stderr, "\nCan't allocate memory in dialog_checklist().\n");
	exit (-1);
    }
    do {
        ret = dialog_checklist (t, av[2], atoi (av[3]), atoi (av[4]),
	    atoi (av[5]), item_no, av + 6, FLAG_RADIO, status);
        if (helptag && ret == 2) {
            dialog_clear();
            dialog_textbox("Help", helptag, 18, 70);
            dialog_clear();
        }
    } while (ret == 2);
    if (ret == 0) {
	for (i = 0; i < item_no; i++)
	    if (status[i])
		fprintf (stderr, "%s", (av + 6)[i * 3]);
    }
    free (status);
    return ret;
}

int
j_inputbox (const char *t, int ac, const char * const * av)
{
    int ret;
    do {
        ret = dialog_inputbox (t, av[2], atoi (av[3]), atoi (av[4]),
			        ac == 6 ? av[5] : (char *) NULL);
        if (helptag && ret == 2) {
            dialog_clear();
            dialog_textbox("Help", helptag, 18, 70);
            dialog_clear();
        }
    } while (ret == 2);
    if (ret == 0)
	fprintf(stderr, dialog_input_result);
    return ret;
}

#ifdef HAVE_GAUGE
int
j_gauge (const char *t, int ac, const char * const * av)
{
    return dialog_gauge (t, av[2], atoi (av[3]), atoi (av[4]),
			 atoi (av[5]));
}
#endif
